### DISCUSSION TITLE
Executing multiple bash functions in a script?

### DISCUSSION THREAD

**Discussion by xfzv** [junegunn/fzf#3148] [Q&A] [answered]
I have the following script:

```bash
#! /usr/bin/env bash

my_function_1() { printf "Function 1\n"; }
my_function_2() { printf "Function 2\n"; }

fzf_menu() {
    local CHOICE="$(fzf --multi < <(
        printf "Function 1\n"
        printf "Function 2\n"
        printf "Exit\n"
    ))"

    case "$CHOICE" in
        "Function 1") my_function_1 ;;
        "Function 2") my_function_2 ;;
        "Exit" | *)
            clear
            exit 0
            ;;
    esac
}

while true; do
    clear
    fzf_menu "$@"
    printf "\n\n\t\t\tPress any key to continue"
    read -r -n 1
done
```

If I select only one entry (either `Function 1` or `Function 2`): all good.

However, if I select both entries with TAB key and press Enter, the script ends with exit code `0.` I guess that's because `$CHOICE` is neither equal to `Function 1` nor `Function 2` in this case so it's considered as `*)`?

I would like both `my_function_1` and `my_function_2` to be executed (one after the other). Is this feasible?

Thanks.

**Comment by xfzv** [comment, marked as answer]
Anyone?

**Reply by Ioulan** [reply]
Hello, I'm not good at bash, so Instead I asked Chat GPT, and suggested me to use IFS (Internal Field Separator) variable to split the selected options into an array then loop through it to execute the corresponding function. 
```bash
#! /usr/bin/env bash

my_function_1() { printf "Function 1.\n"; }
my_function_2() { printf "Function 2.\n"; }

fzf_menu() {
  local choice
  choice="$(printf "Function 1\nFunction 2\nExit\n" | fzf --multi --select-1)"

  IFS=$'\n' read -rd '' -a choices <<< "$choice"
  for c in "${choices[@]}"; do
    case "$c" in
      "Function 1") my_function_1 ;;
      "Function 2") my_function_2 ;;
      "Exit") 
        clear
        exit 0
        ;;
    esac
  done
}

while true; do
  clear
  fzf_menu "$@"
  printf "\n\n\t\t\tPress any key to continue"
  read -r -n 1
done
```
According also to the AI (wherein the first snippet is the one you provided, and the second one is the snippet that I give above:
> The difference between the two code snippets is that in the first one, fzf is reading its input from a process substitution. This causes the $CHOICE variable to contain the selected options separated by newlines, whereas in the second code snippet, fzf is reading its input from a pipe, causing the $choice variable to contain all the selected options in a single line separated by newlines.

Diff
```diff
#! /usr/bin/env bash

-my_function_1() { printf "Function 1\n"; }
+my_function_1() { printf "Function 1.\n"; }
-my_function_2() { printf "Function 2\n"; }
+my_function_1() { printf "Function 2.\n"; }

fzf_menu() {
-    local CHOICE="$(fzf --multi < <(
-        printf "Function 1\n"
-        printf "Function 2\n"
-        printf "Exit\n"
-    ))"
# Linter suggestion: Declare and assign separately to avoid masking return values.
+  local choice
+  choice="$(printf "Function 1\nFunction 2\nExit\n" | fzf --multi --select-1)"

-    case "$CHOICE" in
-        "Function 1") my_function_1 ;;
-        "Function 2") my_function_2 ;;
-        "Exit" | *)
-            clear
-            exit 0
-            ;;
+  IFS=$'\n' read -rd '' -a choices <<< "$choice"
+  for c in "${choices[@]}"; do
+    case "$c" in
+      "Function 1") my_function_1 ;;
+      "Function 2") my_function_2 ;;
+      "Exit") 
+        clear
+        exit 0
+        ;;
    esac
+ done
}

while true; do
    clear
    fzf_menu "$@"
    printf "\n\n\t\t\tPress any key to continue"
    read -r -n 1
done
```

**Reply by xfzv** [reply]
> Hello, I'm not good at bash, so Instead I asked Chat GPT, and suggested me to use IFS (Internal Field Separator) variable to split the selected options into an array then loop through it to execute the corresponding function.


Thank you very much, works perfectly!
